homeassistant/components/pglab/discovery.py

```diff
@@ -0,0 +1,277 @@
+"""Discovery a PG LAB Electronics devices."""
+
+from collections.abc import Callable
+from dataclasses import dataclass
+import json
+from typing import Any
+
+from pypglab.device import Device as PyPGLabDevice
+from pypglab.mqtt import Client as PyPGLabMqttClient
+
+from homeassistant.components.mqtt import ReceiveMessage
+from homeassistant.components.mqtt.subscription import (
+    EntitySubscription,
+    async_prepare_subscribe_topics,
+    async_subscribe_topics,
+    async_unsubscribe_topics,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import Platform
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers import device_registry as dr, entity_registry as er
+from homeassistant.helpers.device_registry import CONNECTION_NETWORK_MAC
+from homeassistant.helpers.dispatcher import (
+    async_dispatcher_connect,
+    async_dispatcher_send,
+)
+from homeassistant.helpers.entity import Entity
+
+from .const import _LOGGER, DISCOVERY_TOPIC, DOMAIN
+
+# Supported platforms
+PLATFORMS = [
+    Platform.SWITCH,
+]
+
+# Used to create a new component entity
+CREATE_NEW_ENTITY = {
+    Platform.SWITCH: "pglab_create_new_entity_switch",
+}
+
+# Define custom PG LAB Config entry
+PGLABConfigEntry = ConfigEntry["PGLabDiscovery"]
+
+
+def get_device_id_from_discovery_topic(topic: str) -> str | None:
+    """From the discovery topic get the PG LAB Electronics device id."""
+
+    # the discovery topi has the following format "pglab/discovery/[Device ID]/config"
+    split_topic = topic.split("/")
+
+    # do a sanity check on the string
+    if len(split_topic) != 4:
+        return None
+
+    if split_topic[3] != "config":
+        return None
+
+    return split_topic[2]
+
+
+class DiscoverDeviceInfo:
+    """It's keeping information of the PGLAB discovered device."""
+
+    def __init__(self, pglab_device: PyPGLabDevice) -> None:
+        """Initialize the device discovery info."""
+
+        # hash string that represent the device actual configuration,
+        # it depends by the number of available relay and shutter
+        # When changed the device enties must be rebuilt
+        self._hash = pglab_device.hash
+        self._entities: list[tuple[str, str]] = []
+
+    def add_entity(self, entity: Entity) -> None:
+        """Add the entity id."""
+        if entity.unique_id:
```

Comment:
Not really sure why I put unique_id check... Entity is actually a HA class and it always has the unique id attribute. Removing the check. Thank you for spotting that.

Reply:
If everything has an unique_id, we should rather `assert entity.unique_id` instead of making it optional, as it can raise the question Erik mentioned when reading the code, which could mislead you

Reply:
joost is right, I added a suggested code change